File: scripts/make_matrix.py
#!/usr/bin/env python3

"""Make matrix from modes"""

import argparse
import glob
import os
import sys
from math import log
import pandas as pd
from scipy.spatial.distance import pdist, squareform

# --------------------------------------------------
def get_args():
    """get args"""
    parser = argparse.ArgumentParser(
        description='Make matrix from modes',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument(
        '-d', '--distance_method', help='Distance method',
        metavar='STR', type=str, default="euclidean")

    parser.add_argument(
        '-m', '--mode_dir', help='Mode directory',
        metavar='DIR', type=str, required=True)

    parser.add_argument(
        '-o', '--out_dir', help='Matrix output dir',
        metavar='DIR', type=str, default=os.getcwd())

    return parser.parse_args()

# --------------------------------------------------
def main():
    """main"""
    args = get_args()
    mode_dir = args.mode_dir
    out_dir = args.out_dir
    distance_method = args.distance_method
    valid_distance = set("""
        braycurtis canberra chebyshev cityblock correlation cosine 
        dice euclidean hamming jaccard kulsinski matching rogerstanimoto 
        russellrao seuclidean sokalmichener sokalsneath sqeuclidean yule.
        """.split())

    if not mode_dir:
        print('--mode_dir is required')
        sys.exit(1)

    if not os.path.isdir(mode_dir):
        print('Bad --mode_dir "{}"'.format(mode_dir))
        sys.exit(1)

    if not distance_method in valid_distance:
        print('--distance_method "{}" is not valid'.format(distance_method))
        print('Please select from {}'.format(', '.join(valid_distance)))
        sys.exit(1)

    if not os.path.isdir(out_dir):
        os.makedirs(out_dir)

    mode_files = list(filter(os.path.isfile,
                             glob.iglob(mode_dir + '/**', recursive=True)))
    print('Found {} mode files'.format(len(mode_files)))

    counts = {}
    for file in mode_files:
        sample1 = os.path.basename(os.path.dirname(file))
        sample2 = os.path.basename(file)
        num = open(file).read().strip()
        if not sample1 in counts:
            counts[sample1] = {}
        counts[sample1][sample2] = int(num)

    all_keys = set(counts.keys())
    for key in all_keys:
        map(all_keys.add, counts[key].keys())

    all_samples = sorted(all_keys)

    raw_file = os.path.join(out_dir, 'matrix_raw.txt')
    raw_fh = open(raw_file, 'w')

    avg_file = os.path.join(out_dir, 'matrix_avg.txt')
    avg_fh = open(avg_file, 'w')

    log_avg_file = os.path.join(out_dir, 'matrix_log_avg.txt')
    log_avg_fh = open(log_avg_file, 'w')

    raw_fh.write('\t'.join([''] + all_samples) + '\n')
    avg_fh.write('\t'.join([''] + all_samples) + '\n')
    log_avg_fh.write('\t'.join([''] + all_samples) + '\n')

    for sample1 in all_samples:
        raw = [sample1]
        avgs = [sample1]
        log_avg = [sample1]

        for sample2 in all_samples:
            n1 = counts[sample1].get(sample2, 0)
            n2 = counts[sample2].get(sample1, 0)
            avg = (n1 + n2) / 2

            raw.append(str(n1))
            avgs.append('{:.0f}'.format(avg))
            log_avg.append('{:.4f}'.format(log(avg)) if avg > 0 else '0')

        raw_fh.write('\t'.join(raw) + '\n')
        avg_fh.write('\t'.join(avgs) + '\n')
        log_avg_fh.write('\t'.join(log_avg) + '\n')

    raw_fh.close()
    avg_fh.close()
    log_avg_fh.close()
    
    dat = pd.read_csv(avg_file, sep="\t", index_col=0)
    dist = pd.DataFrame(squareform(pdist(dat.values, distance_method)),
                        index=dat.index,
                        columns=dat.index)

    norm_file = os.path.join(out_dir, 'matrix_normalized.txt')
    dist.to_csv(norm_file, sep='\t')

    print('Done, see files in dir "{}"'.format(out_dir))

# --------------------------------------------------
if __name__ == '__main__':
    main()

File: scripts/make_matrix.r
#!/usr/bin/env Rscript

suppressMessages(library("optparse"))
suppressMessages(library("R.utils"))

option_list = list (
  make_option(c("-m", "--mode_dir"), 
              type = "character", 
              default = "",
              help = "Mode directory", 
              metavar="character"
  ),
  make_option(c("-o", "--out_dir"), 
              type = "character", 
              default = '',
              help = "Output directory (--file dir)"
  ),
  make_option(c("-n", "--read_nums"), 
              type = "character", 
              default = "",
              help = "CSV file with number of reads for each file"
  )
)

opt_parser = OptionParser(option_list = option_list)
opt        = parse_args(opt_parser)
out.dir    = opt$out_dir
mode.dir   = opt$mode_dir
read.nums  = opt$read_nums

if (nchar(mode.dir) == 0) {
  stop("--mode_dir is required")
}

if (!dir.exists(mode.dir)) {
  stop(paste("Bad --mode_dir", matrix_file))
}

if (nchar(read.nums) == 0) {
  stop("--read_nums is required")
}

if (!file.exists(read.nums)) {
  stop("--read_nums must be a file")
}

if (nchar(out.dir) == 0) {
  out.dir = mode.dir
}

if (!dir.exists(out.dir)) {
  dir.create(out.dir)
}

if (!dir.exists(out.dir)) {
  dir.create(out.dir)
}

# mode.dir = "~/work/dolphin/mode/"
# read.nums = "~/work/dolphin/counts.csv"

num.reads = read.csv(read.nums, header=FALSE)
colnames(num.reads) = c('file', 'num')
files = list.files(path = mode.dir, recursive = T, full.names = T)
printf("Processing %s files in '%s'\n", length(files), mode.dir)

samples = sort(unlist(unique(Map(basename, files))))
num.samples = length(samples)
df = data.frame(matrix(data = 0, 
                       nrow = num.samples, 
                       ncol = num.samples))
colnames(df) = samples
rownames(df) = samples

norm.df = df[0,]
for (path in files) {
  s1 = basename(dirname(path)) # the index
  s2 = basename(path) # the query
  n = as.integer(readLines(path))
  s2.reads = num.reads[num.reads$file == s2, 'num']
  df[s1, s2] = n # raw
  norm.df[s1, s2] = n/s2.reads # normalized by the size of the query 
}

# Clone the data.frame structure (no data)
# Fill in with mean of A-B/B-A
avg.df = norm.df[0,]
for (c in colnames(norm.df)) {
  for (r in rownames(norm.df)) {
    # log of a number < 1 is negative, so need to mult/div by 100 (percentage)
    avg.df[r, c] = log(mean(c(norm.df[r,c], norm.df[c,r])) * 100) / 100
  }
}

write.table(df, file = file.path(out.dir, "matrix_raw.tab"))
write.table(norm.df, file = file.path(out.dir, "matrix_norm.tab"))
write.table(avg.df, file = file.path(out.dir, "matrix_norm_log_avg.tab"))
write.table(1 - avg.df, file = file.path(out.dir, "matrix_dist.tab"))

printf("Done, wrote raw/norm files to out_dir '%s'\n", out.dir)
